src/napi-inl.h

```diff
@@ -0,0 +1,1724 @@
+  return Type() == napi_boolean;
+}
+
+inline bool Value::IsNumber() const {
+  return Type() == napi_number;
+}
+
+inline bool Value::IsString() const {
+  return Type() == napi_string;
+}
+
+inline bool Value::IsSymbol() const {
+  return Type() == napi_symbol;
+}
+
+inline bool Value::IsArray() const {
+  if (_value == nullptr) {
+    return false;
+  }
+
+  bool result;
+  napi_status status = napi_is_array(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return result;
+}
+
+inline bool Value::IsArrayBuffer() const {
+  if (_value == nullptr) {
+    return false;
+  }
+
+  bool result;
+  napi_status status = napi_is_arraybuffer(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return result;
+}
+
+inline bool Value::IsTypedArray() const {
+  if (_value == nullptr) {
+    return false;
+  }
+
+  bool result;
+  napi_status status = napi_is_typedarray(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return result;
+}
+
+inline bool Value::IsObject() const {
+  return Type() == napi_object;
+}
+
+inline bool Value::IsFunction() const {
+  return Type() == napi_function;
+}
+
+inline bool Value::IsBuffer() const {
+  if (_value == nullptr) {
+    return false;
+  }
+
+  bool result;
+  napi_status status = napi_buffer_has_instance(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return result;
+}
+
+inline Boolean Value::AsBoolean() const {
+  return Boolean(_env, _value);
+}
+
+inline Number Value::AsNumber() const {
+  return Number(_env, _value);
+}
+
+inline String Value::AsString() const {
+  return String(_env, _value);
+}
+
+inline Object Value::AsObject() const {
+  return Object(_env, _value);
+}
+
+inline Array Value::AsArray() const {
+  return Array(_env, _value);
+}
+
+inline TypedArray Value::AsTypedArray() const {
+  return TypedArray(_env, _value);
+}
+
+inline Function Value::AsFunction() const {
+  return Function(_env, _value);
+}
+
+inline Buffer Value::AsBuffer() const {
+  return Buffer(_env, _value);
+}
+
+inline Boolean Value::ToBoolean() const {
+  napi_value result;
+  napi_status status = napi_coerce_to_bool(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return Boolean(_env, result);
+}
+
+inline Number Value::ToNumber() const {
+  napi_value result;
+  napi_status status = napi_coerce_to_number(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return Number(_env, result);
+}
+
+inline String Value::ToString() const {
+  napi_value result;
+  napi_status status = napi_coerce_to_string(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return String(_env, result);
+}
+
+inline Object Value::ToObject() const {
+  napi_value result;
+  napi_status status = napi_coerce_to_object(_env, _value, &result);
+  if (status != napi_ok) throw Error::New(Env());
+  return Object(_env, result);
+}
```

Comment:
1. I'd have to use template specialization to call the different `napi_coerce_*` functions for each implementation, so the code wouldn't be any simpler with templates.
2. This is more similar to V8, which has separate ToSomething() methods for each supported type.